### ISSUE TITLE
Advanced Auditing 1.12 umbrella bug

### ISSUE THREAD

**Issue by x13n** [kubernetes/kubernetes#65266]
This is a continuation of https://github.com/kubernetes/kubernetes/issues/60392

## API-related changes
- [x] Graduate API to stable #65891
- [x] Integration with admission
    - [x] Annotate audit logs in remote admission webhooks
@CaoShuFeng is working on this in #58679
- [x] Add user-agent to audit events #64791

## Bugfixes and improvements
- [x] Completely remove legacy audit logging https://github.com/kubernetes/kubernetes/pull/65862
- [ ] [Optional] [Always retry audit webhook](https://github.com/kubernetes/kubernetes/issues/55621)
- [ ] [Optional] [Dynamic audit configuration](https://github.com/kubernetes/community/pull/2188)

## Policy changes
- [ ] Add a setting in audit policy that will allow rejecting API request when audit logging fails https://github.com/kubernetes/kubernetes/pull/65763
- [ ] [Optional] [GCE audit policy should be made re-usable by other setups](https://github.com/kubernetes/kubernetes/issues/53321)
- [ ] [Optional] Audit policy should be tested (e.g. that it includes all core resources, [more context](https://github.com/kubernetes/kubernetes/issues/52265#issuecomment-329243972))

## To discuss
- [ ] [Optional] Auditing federation setups

I mostly carried over the unfinished work from 1.11 issue. Two things added: the ability to reject apiserver requests when audit logging fails (configurable via audit policy) and - optionally - recent proposal for dynamic audit configuration.

/kind feature
/sig auth
/area audit

cc @loburm @tallclair @soltysh

**Comment by CaoShuFeng** [comment]
Maybe add this: https://github.com/kubernetes/kubernetes/issues/64791

**Comment by piosz** [comment]
cc @destijl

**Comment by tallclair** [comment]
> Add a setting in audit policy that will allow rejecting API request when audit logging fails

We already have that per-backend with `Blocking` mode. Is the idea that it would be nice to require blocking for only some highly-sensitive requests?

**Comment by x13n** [comment]
@CaoShuFeng I see Tim already added this.
@tallclair As far as I can tell, blocking will wait for audit event to be written, but upon a write failure (say, the webhook goes down, or the disk used for log backend is full) the request to API server will be processed anyway. The audit failure will only be written to the logs. I think that either:
- 'blocking' should mean that the request to API server fails if audit logging fails OR
- there should be a third option ('blocking-strict'? naming is hard) that would do that and existing 'blocking' will continue to behave as today.

**Comment by tallclair** [comment]
I see, thanks for clarifying. Yeah, blocking should have been that (blocking-strict sgtm)

**Comment by x13n** [comment]
Sorry, can you clarify which one do you think makes more sense? Adding third option (e.g. blocking-strict) or changing behavior of existing 'blocking' setting?

**Comment by loburm** [comment]
@timstclair could you approve it for milestone?

**Comment by tallclair** [comment]
I don't think we can change `blocking` without breaking backwards compatibliity, so I'm in favor of `blocking-strict`

**Comment by liggitt** [comment]
> I don't think we can change blocking without breaking backwards compatibliity, so I'm in favor of blocking-strict

what would it mean for a mutating request to fail if the responsecomplete audit logging fails after the object is persisted? just a failure http status?

**Comment by tallclair** [comment]
Right, to clarify this would only work for the `RequestReceived` stage. Actually, that could be used to get this behavior only for sensitive requests (by omitting RequestReceived for less sensitive requests).

**Comment by x13n** [comment]
Yeah, this makes sense only for `RequestReceived` stage, other stages would continue to stay best effort.

**Comment by loburm** [comment]
@CaoShuFeng I see that you have been performing API graduation from alpha to beta last time. Do you want to take care about doing the same for stable, or should I take it?

And do you think that there are any blockers for it?

**Comment by CaoShuFeng** [comment]
> Do you want to take care about doing the same for stable, or should I take it?

Would be good either way.

I remembered that we will delete some field whens upgrade the audit API.
https://github.com/kubernetes/kubernetes/issues/52160#issue-256206047


> And do you think that there are any blockers for it?

Not very sure. Do we wait until [issues here](https://github.com/kubernetes/kubernetes/issues/65266#issue-333995465) get solved?

**Comment by loburm** [comment]
> I remembered that we will delete some fields when upgrade the audit API.
#52160 (comment)
Thanks for pointing. Do we need any period for testing newly introduced/renamed fields?

> Not very sure. Do we wait until issues here get solved?
user-agent field was added. @x13n could you update it?
Other changes (non optional one) are not changing API fields. Code removal for legacy audit logging can be done in parallel.

**Comment by CaoShuFeng** [comment]
> Do we need any period for testing newly introduced/renamed fields?

Agree. I think so.

> Other changes (non optional one) are not changing API fields. Code removal for legacy audit logging can be done in parallel.

Sounds good to me.  I could start coding in Friday, if I am responsible for upgrading the API. What do you think?

**Comment by loburm** [comment]
> Sounds good to me. I could start coding in Friday, if I am responsible for upgrading the API. What do you think?

Sounds as a plan. In meantime, I'll start working on removing legacy audit logging.

**Comment by soltysh** [comment]
> Completely remove legacy audit logging

@x13n what do you mean by that? When was it deprecated?

**Comment by x13n** [comment]
@soltysh This one was a carryover from https://github.com/kubernetes/kubernetes/issues/60392. If it wasn't deprecated then I agree we should deprecate it first.

@loburm FYI

**Comment by soltysh** [comment]
I haven't heard it being deprecated, but I might be wrong. If it wasn't please update this to deprecate it for now and not remove.

**Comment by loburm** [comment]
@soltysh 
It was deprecated according to the official documentation: https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#legacy-audit

But also I have checked release notes from 1.8 release, and looks like this information was not added there. Not sure what is going to be the correct way to proceed now.

**Comment by soltysh** [comment]
> It was deprecated according to the official documentation: https://kubernetes.io/docs/tasks/debug-application-cluster/audit/#legacy-audit

I missed that or forgot about it :wink: in that case, we're good to completely remove it, especially that it was disabled by default since 1.8.  Just ensure to have a proper message that we are removing it in release notes, this time.

**Comment by guineveresaenger** [comment]
Please continue this work in 1.13.
/milestone clear
/milestone v1.13

**Comment by liggitt** [comment]
Actually, we'll tombstone this to track what was done in 1.12 and copy unfinished items to a new issue

/milestone v1.12
/close
